crypto/fipsmodule/cipher/e_aesccm.c

```diff
@@ -456,3 +494,239 @@ DEFINE_METHOD_FUNCTION(EVP_AEAD, EVP_aead_aes_128_ccm_matter) {
+        return len;
+      } else {
+        return -1;
+      }
+    }
+  }
+
+  // The tag must be set before decrypting any data.
+  if (!EVP_CIPHER_CTX_encrypting(ctx) && !cipher_ctx->tag_set) {
+    return -1;
+  }
+  if (!cipher_ctx->len_set) {
+    return -1;
+  }
+  if (!cipher_ctx->ccm_set) {
+    // Initialize the ccm_state if this did not happen during the AAD update.
+    if (!ccm128_init_state(ccm_ctx, ccm_state, &cipher_ctx->ks.ks,
+                           cipher_ctx->nonce, CCM_L_TO_NONCE_LEN(cipher_ctx->L),
+                           NULL, 0, cipher_ctx->message_len)) {
+      return -1;
+    }
+    cipher_ctx->ccm_set = 1;
+  }
+
+  if (EVP_CIPHER_CTX_encrypting(ctx)) {
+    // Encryption path. Compute CBC-MAC on plaintext and then encrypt.
+    if (!ccm128_compute_mac(ccm_ctx, ccm_state, &cipher_ctx->ks.ks,
+                            cipher_ctx->tag, cipher_ctx->M, in, len)) {
+      return -1;
+    }
+    if (!ccm128_encrypt(ccm_ctx, ccm_state, &cipher_ctx->ks.ks, out, in, len)) {
+      return -1;
+    }
+    cipher_ctx->tag_set = 1;
+  } else {
+    // Decryption path. Compute the plaintext then compute its CBC-MAC.
+    // n.b. The method says encrypt, but it works both ways.
+    if (!ccm128_encrypt(ccm_ctx, ccm_state, &cipher_ctx->ks.ks, out, in, len)) {
+      return -1;
+    }
+    uint8_t computed_tag[EVP_AEAD_AES_CCM_MAX_TAG_LEN] = {0};
+    if (!ccm128_compute_mac(ccm_ctx, ccm_state, &cipher_ctx->ks.ks,
+                            computed_tag, cipher_ctx->M, out, len)) {
+      return -1;
+    }
+    // Validate the tag and invalidate the output if it doesn't match.
+    if (OPENSSL_memcmp(cipher_ctx->tag, computed_tag, cipher_ctx->M)) {
+      OPENSSL_cleanse(out, len);
+      return -1;
+    }
+    cipher_ctx->iv_set = 0;
+    cipher_ctx->tag_set = 0;
+    cipher_ctx->len_set = 0;
+    cipher_ctx->ccm_set = 0;
+  }
+  return (int) len;
+}
+
+static void cipher_aes_ccm_cleanup(EVP_CIPHER_CTX *ctx) {}
+
+static int cipher_aes_ccm_ctrl_set_L(CIPHER_AES_CCM_CTX *ctx, int L) {
+  if (L < 2 || L > 8) {
+    return 0;
+  }
+  ctx->L = L;
+  return 1;
+}
+
+static int cipher_aes_ccm_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg,
+                               void *ptr) {
+  CIPHER_AES_CCM_CTX *cipher_ctx = CCM_CTX(ctx);
+  switch (type) {
+    case EVP_CTRL_INIT:
+      OPENSSL_cleanse(cipher_ctx, sizeof(CIPHER_AES_CCM_CTX));
+      cipher_ctx->key_set = 0;
+      cipher_ctx->iv_set = 0;
+      cipher_ctx->tag_set = 0;
+      cipher_ctx->len_set = 0;
+      cipher_ctx->ccm_set = 0;
+      cipher_ctx->L = 8;
+      cipher_ctx->M = 14;
+      cipher_ctx->message_len = 0;
+      return 1;
+    case EVP_CTRL_GET_IVLEN:
+      *(uint32_t *)ptr = CCM_L_TO_NONCE_LEN(cipher_ctx->L);
+      return 1;
+    case EVP_CTRL_AEAD_SET_IVLEN:
+      // The nonce (IV) length is 15-L, compute L here and set it below to "set"
+      // the IV length.
+      return cipher_aes_ccm_ctrl_set_L(cipher_ctx, 15 - arg);
```

Comment:
```suggestion
      return cipher_aes_ccm_ctrl_set_L(cipher_ctx, CCM_L_TO_NONCE_LEN(arg));
```

Reply:
Super pedantic, but this was an intentional choice. Though this works algebraically, this is not technically correct. I named the macro `CCM_L_TO_NONCE_LEN` and this particular line calls for converting a nonce length to L.